qpid-jms-client/src/main/java/org/apache/qpid/jms/JmsSession.java

```diff
@@ -194,6 +198,48 @@ public void onPendingFailure(ProviderException cause) {
         }
     }
 
+    private void processCompletions() {
+        assert processCompletion.get();
+        completionThread = Thread.currentThread();
+        try {
+            final Runnable completionTask = completionTasks.poll();
+            if (completionTask != null) {
+                try {
+                    completionTask.run();
+                } catch (Throwable t) {
+                    LOG.debug("errored on processCompletions duty cycle", t);
+                }
+            }
+        } finally {
+            completionThread = null;
+            processCompletion.set(false);
+        }
+        if (completionTasks.isEmpty()) {
+            return;
+        }
+        // a racing asyncProcessCompletion has won: no need to fire a continuation
+        if (!processCompletion.compareAndSet(false, true)) {
+            return;
+        }
+        getCompletionExecutor().execute(this::processCompletions);
+    }
+
+    private void asyncProcessCompletion(final Runnable completionTask) {
+        asyncProcessCompletion(completionTask, false);
+    }
+
+    private void asyncProcessCompletion(final Runnable completionTask, final boolean ignoreSessionClosed) {
+        if (!ignoreSessionClosed) {
```

Comment:
I can let them to pass, but this one was to mimic the original reject handler installed for the completion single threaded executor: I admit I didn't put much thoughts on this to validate if it can be saved

Reply:
Ill admit I forgot about that bit, but yes it is a little different - that is something that only occurs later, after the executor is shut down in the existing non-shared-completion case, and is handling a case that probably wont occur, and also shouldnt matter if it did because something else already cleaned up anything related to it before killing the executor.

Here, the work would start being thrown away immediatelym even before the producer itself is necessarily notified its being considered closed, and long before the executor is being shut down (which in the regular non-shared-pool case would still happen), so it could actually be a quite significant change in behaviour. I think this introduces a hole that didnt exist before.